src/common/perf_counters_key.h

```diff
@@ -63,7 +63,11 @@ class label_iterator {
   reference operator*() const { return state->label; }
   pointer operator->() const { return &state->label; }
 
-  auto operator<=>(const label_iterator& rhs) const = default;
+  // Keep equality; defaulting this requires only operator== on iterator_state
+  bool operator==(const label_iterator& rhs) const = default;
+
+  // Disable three-way comparison since iterator_state doesn't support it
+  auto operator<=>(const label_iterator& rhs) const = delete;
```

Comment:
can you help me understand what in `iterator_state` doesn't support this? it has its own defaulted `operator<=>`, and all  of the primitive types involved (optional, pair, string_view) provide spaceship operators too

what compiler version is complaining about this?

Reply:
thanks, i found the log in https://jenkins.ceph.com/job/ceph-windows-pull-requests/64686/

> /home/ubuntu/ceph/src/common/perf_counters_key.h:66:8: error: explicitly defaulted three-way comparison operator is implicitly deleted [-Werror,-Wdefaulted-function-deleted]

crazy that this was okay without -Werror. regardless, i would argue that -Werror is not helping to improve code quality in this case

Reply:
> As for backwards compatibility, a change was made recently to set -std=c++23. That might cause many more headaches :)

howso?

Reply:
there was some discussion about the switch to c++23 on the mailing list in https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/L4GKNZUYTIBTLYNLM4PZ54NPAV4W4IHX/ if you're interested. it turned out that gcc-11 was good enough to enable c++23 on main, though it certainly will constrain which new language features we'll be able to use

Reply:
> How would that be caught before committing it to main?

teuthology testing is a critical part of the qa process. the first step of that is to produce package/container builds for each supported distro using their native toolchain. so a pull request that tries to use 'deducing this' would fail qa and hopefully not be merged

Ilya raised this concern with his approval of https://github.com/ceph/ceph/pull/64426#pullrequestreview-3009897556, and it's a valid one. process mistakes that break the build on main do happen occassionally, but get caught/addressed quickly